hibernate-rx-core/src/main/java/org/hibernate/rx/loader/RxOuterJoinLoader.java

```diff
@@ -72,9 +73,18 @@ public RxOuterJoinLoader(SessionFactoryImplementor factory, LoadQueryInfluencers
 		return doRxQuery( session, queryParameters, returnProxies, forcedResultTransformer )
 				.handle( (list, e) -> {
 					persistenceContext.afterLoad();
-					if (e == null) {
-						persistenceContext.initializeNonLazyCollections();
+					if (e != null) {
+						persistenceContext.setDefaultReadOnly(defaultReadOnlyOrig);
+						RxUtil.rethrow(e);
 					}
+					return list;
+				})
+				.thenCompose(list -> {
+					// only initialize non-lazy collections after everything else has been refreshed
+					return ((RxPersistenceContextAdapter) persistenceContext ).rxInitializeNonLazyCollections()
+							.thenApply(v -> list);
+				})
+				.handle( (list, e) -> {
```

Comment:
I imagine there's a more elegant way to do this than `handle( ... ).thenCompose( ... ).handle( ... )`.

Suggestions?

Reply:
My suggestions:

1. Can you suck the calls to `setDefaultReadOnly()` into `rxInitializeNonLazyCollections()`? They seem be called around every invocation? Then you wouldn't need `handle()` at all, if I'm not mistaken.
2. For single-line anonymous functions I would prefer not using `{ return .... }`. Minor thing.

Reply:
@gavinking, 

> * Can you suck the calls to `setDefaultReadOnly()` into `rxInitializeNonLazyCollections()`? They seem be called around every invocation? Then you wouldn't need `handle()` at all, if I'm not mistaken.

Related to your question -- I'm not sure if the temporary setting above (out of view), in line 65, is in the proper place. Could this setting inadvertently affect future operations, since the setting only gets restored after `#doRxQuery` and friends complete?

I think it would be better to move this temporary "set" and "unset" to the code where objects are actually being added to the `PersistenceContext`.

I think this may also affect `persistenceContext.beforeLoad()` and `persistenceContext.afterLoad()`.

Am I understanding this correctly?  If so, I would prefer to deal with this in a separate issue.

> * For single-line anonymous functions I would prefer not using `{ return .... }`. Minor thing.

I pushed a commit with that change.

BTW, when this gets approved, I'll squash the commits, and make sure the message refers to #95.